02-express-tutorial/app.js

```diff
@@ -1 +1,97 @@
-console.log('Express Tutorial')
+const express = require('express');
+const app = express();
+const { products } = require("./data")
+
+
+/**
+ * Create HTTP server
+ */
+// const http = require('http')
+const port = 3000;
+// app.set('port', port)
+// const server = http.createServer(app)
+// server.listen(port, () => {
+//     console.log(`app listening at http://localhost:${port}`)
+// })
+
+app.use(express.static("./public"))
+
+// First test endpoint
+app.get('/api/v1/test', (request, response) => {
+    response.json({
+        message: 'My test API endpoint!'
+    })
+})
+
+// Get all products
+app.get('/api/v1/products', (request, response) => {
+    response.json(products)
+})
+
+// Get product by Id
+app.get('/api/v1/products/:productId', (request, response) => {
+    
+    const productId = parseInt(request.params.productId)
+    const product = products.find( p => p.id === productId )
+    
+    if( !product ) {
+        response.status(400).json({
+            message: 'Product not found!'
+        })
+    }
+
+    response.json(product)
+})
+
+// Search product query
+app.get('/api/v1/query', (request, response) => {
+    const defaultLimit = 5;
+
+    const search = request.query.search && request.query.search.toLocaleLowerCase() || ""
+    const limit = Math.abs( parseInt(request.query.limit) ) || defaultLimit
+    const minPrice = parseFloat(request.query.minPrice) || -1
+    const maxPrice = parseFloat(request.query.maxPrice) || -1
+
+    // Handle error: Always minPrice < maxPrice
+    if( minPrice!==-1 && maxPrice!==-1 && minPrice>maxPrice ) {
+        response.status(400).json({
+            message: "Minimum and maximum price conflict!"
+        })
+    }
+    
+    let resultProducts = []
+    // Check if at least one query param provided to filter the result
+    if( search !== "" || minPrice !== -1 || maxPrice !== -1 ) {
+
+        resultProducts = products.filter(
+            (p) => {
+                const searchResult = search !== "" ? (p.name.toLowerCase().indexOf(search) >= 0) : true
+                const minPriceResult = minPrice !== -1 ? (p.price >= minPrice) : true
+                const maxPriceResult = maxPrice !== -1 ? (p.price <= maxPrice) : true
```

Comment:
The ternary conditional statements are repetitive with the conditionals on line 64.  There is a coding principle known as "Do Not Repeat [Yourself]", or "DRY".

Reference: https://www.geeksforgeeks.org/dont-repeat-yourselfdry-in-software-development/